### PR TITLE
Decouple inference from training and introduce callbacks

### PR DESCRIPTION
**Pull request by workofart**
## Description
Right now, LLMTrainer is highly coupled with inference. This is problematic if we want to extend the trainer or use the trainer for different language modeling tasks like supervised fine-tuning.

This PR will decouple the inference part of the evaluation into a callback. Then it's the caller's responsibility to actually configure which callbacks to use during evaluation. The vanilla evaluate() will be a simple forward pass + loss computation on the validation data.

## Tests
All unit tests passed.

GPT-2 Sample Run
```
(ml-by-hand) ➜  ml-by-hand git:(cleanup-trainer) ✗ python -m examples.gpt_2
2026-04-14 22:52:52,403 - INFO - 1115394 characters in the entire dataset. Sample:
First Citizen:
Before we proceed any further, hear me speak.

All:
Speak, speak.

First Citizen:
You
2026-04-14 22:52:52,403 - INFO - Loading the vocabulary from disk.
2026-04-14 22:52:52,405 - INFO - Found existing encoded data at 'training_data/bpe_3000_shakespeare_encoded_data.npz', loading it instead of re-encoding.
2026-04-14 22:52:52,407 - INFO - Vocabulary size: 3260
2026-04-14 22:52:52,407 - INFO - Encoded data length: 356417
Data length: len(train_data)=320775 len(test_data)=35642
2026-04-14 22:52:52,411 - INFO - No resume_epoch given; initializing model & optimizer from scratch.
2026-04-14 22:52:52,422 - INFO - Training GPT2 with 45.11M parameters.
Training:   0%|                                           | 0/10 [00:00<?, ?it/s2026-04-14 22:53:08,031 - INFO - Teacher forcing mode on!!
Groundtruth:
 the thing that feeds their fury:
Though little fire grows great with little wind,
Yet extreme gusts will blow out
2026-04-14 22:53:08,031 - INFO - Model:

,, the,,, the,,

 the,,,, the the,,
And the,:,,,
 the,

...

2026-04-14 22:56:07,884 - INFO - [Epoch 9]
grad_l2_norm = 0.8746
val_loss = 4.7662
train_loss = 5.1653
LR = 0.0001
2026-04-14 22:56:07,885 - INFO - Saved training metrics to training_runs/GPT2_shakespeare_mini.npz
2026-04-14 22:56:07,887 - INFO - Model:

Now see it could know being too more than
k? OVERCLLOFutous cause on her children. Gidarels I must learn,
Which, ancriesgar with good time that I must wish, the queen?

GREGincestetho, or eclation of me doth I as he told thou hast bear it.

JULIET:
You know.

BUCKINGHAM:
Fnow,
```

### CODE REVIEW

File: autograd/tools/trainer.py [hunk 1]
```diff
@@ -458,15 +456,7 @@ def evaluate(self, train_data_loader, val_data_loader, epoch) -> float:
...
-            if (
-                self.sample_predictions
-                and epoch
-                % max(
-                    1,
-                    self.config.total_epochs // min(10, self.config.total_epochs),
-                )
-                == 0
-            ):
+            if self.sample_predictions:
```
Comment (workofart) [inline autograd/tools/trainer.py:459]: Cleaning this up, and just print the predictions if the `sample_predictions` flag is turned on.
That would make the evaluate() function lighter and have less dependencies.

File: autograd/tools/trainer.py [hunk 2]
```diff
@@ -629,63 +635,33 @@ def train_step(self, batch_data, data_loader) -> float:
...
-            batch_count += 1
-        avg_val_loss = total_loss / max(batch_count, 1)
-
-        if epoch % 2 == 0:  # Optionally perform inference every 2 epochs
-            self._perform_inference(train_data_loader)
-
-        self.model.train()
-        return avg_val_loss
-
-    def _perform_inference(self, train_data_loader):
```
Comment (workofart) [inline autograd/tools/trainer.py:660]: These two functions are moved to callback.py
